From: Alan Chen <quic_alache@quicinc.com>
Date: Thu, 17 Feb 2022 13:30:20 -0800
Subject: [PATCH] qcacld-3.0: Use cfg80211_register_netdevice() for kernel
 5.12+

For Kernel 5.12+ in order to be in line with cfg80211 changes
in upstream code, use cfg80211_register_netdevice() and
cfg80211_unregister_netdevice() when request to register device is
coming from add virtual interface.

(partial cherry pick from commit 40a7b4a2f97c7f2dc62944963a72f0f0c64d83c6)

https://onedigi.atlassian.net/browse/DEL-7986

Signed-off-by: Isaac Hermida <isaac.hermida@digi.com>
---
 core/hdd/src/wlan_hdd_main.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/core/hdd/src/wlan_hdd_main.c b/core/hdd/src/wlan_hdd_main.c
index a866a90110f6..235b53cbb94c 100644
--- a/core/hdd/src/wlan_hdd_main.c
+++ b/core/hdd/src/wlan_hdd_main.c
@@ -5577,7 +5577,11 @@ static QDF_STATUS hdd_register_interface(struct hdd_adapter *adapter, bool rtnl_
 			}
 		}
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 12, 0)
+		ret = cfg80211_register_netdevice(dev);
+#else
 		ret = register_netdevice(dev);
+#endif
 		if (ret) {
 			hdd_err("register_netdevice(%s) failed, err = 0x%x",
 				dev->name, ret);
@@ -6289,7 +6293,11 @@ static void hdd_cleanup_adapter(struct hdd_context *hdd_ctx,
 
 	if (test_bit(NET_DEVICE_REGISTERED, &adapter->event_flags)) {
 		if (rtnl_held)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 12, 0)
+			cfg80211_unregister_netdevice(dev);
+#else
 			unregister_netdevice(dev);
+#endif
 		else
 			unregister_netdev(dev);
 		/*
